Source/Core/Common/Image.cpp

```diff
@@ -4,181 +4,37 @@
-{
-  for (png_size_t i = 0; i < row_info->rowbytes; i += 4, data += 4)
-  {
-    const png_byte alpha = data[3];
-    u32 w;
-
-    if (alpha == 0)
-    {
-      w = 0;
-    }
-    else
-    {
-      png_byte red = data[0];
-      png_byte green = data[1];
-      png_byte blue = data[2];
-
-      if (alpha != 0xff)
-      {
-        red = MultiplyAlpha(alpha, red);
-        green = MultiplyAlpha(alpha, green);
-        blue = MultiplyAlpha(alpha, blue);
-      }
-      w = (alpha << 24) | (red << 16) | (green << 8) | (blue << 0);
-    }
-
-    std::memcpy(data, &w, sizeof(w));
-  }
-}
-
-struct ReadProgress
-{
-  const u8* current;
-  const u8* end;
-};
-
-static void ReadCallback(png_structp png, png_bytep data, png_size_t size)
-{
-  ReadProgress* progress = static_cast<ReadProgress*>(png_get_io_ptr(png));
-  for (size_t i = 0; i < size; ++i)
-  {
-    if (progress->current >= progress->end)
-      png_error(png, "Read beyond end of file");  // This makes us longjmp back to LoadPNG
-
-    data[i] = *(progress->current);
-    ++(progress->current);
-  }
-}
-
-static void PNGErrorCallback(png_structp png, png_const_charp error_msg)
-{
-  ERROR_LOG(COMMON, "PNG loading error: %s", error_msg);
-  longjmp(png_jmpbuf(png), 1);
-}
-
-#ifdef _MSC_VER
-#pragma warning(push)
-#pragma warning(disable : 4611)
-// VS shows the following warning even though no C++ objects are destroyed in LoadPNG:
-// "warning C4611: interaction between '_setjmp' and C++ object destruction is non-portable"
-#endif
-
 bool LoadPNG(const std::vector<u8>& input, std::vector<u8>* data_out, u32* width_out,
              u32* height_out)
 {
-  // The const_cast is only required for libpng versions older than 1.5
-  const bool is_png = !png_sig_cmp(const_cast<u8*>(input.data()), 0, input.size());
-  if (!is_png)
-    return false;
+  // Using the 'Simplified API' of libpng; see section V in the libpng manual.
 
-  png_struct* png =
-      png_create_read_struct(PNG_LIBPNG_VER_STRING, nullptr, PNGErrorCallback, nullptr);
-  if (!png)
+  // Read header
+  png_image png = {};
+  png.version = PNG_IMAGE_VERSION;
+  if (!png_image_begin_read_from_memory(&png, &input[0], input.size()))
     return false;
 
-  png_info* info = png_create_info_struct(png);
-  if (!info)
-  {
-    png_destroy_read_struct(&png, &info, nullptr);
-    return false;
-  }
-
-  // It would've been nice to use std::vector for this, but using setjmp safely
-  // means that we have to use this manually managed volatile pointer garbage instead.
-  png_byte** volatile row_pointers_volatile = nullptr;
+  // Prepare output vector
+  size_t png_size = PNG_IMAGE_SIZE(png);
+  data_out->resize(png_size);
 
-  if (setjmp(png_jmpbuf(png)))
-  {
-    free(row_pointers_volatile);
-    png_destroy_read_struct(&png, &info, nullptr);
+  // Convert to RGBA and write into output vector
+  png.format = PNG_FORMAT_RGBA;
+  if (!png_image_finish_read(&png, nullptr, &(*data_out)[0], 0, nullptr))
```

Comment:
noticing this now, sorry, but wouldn't it be simpler/more readable to do `data_out->data()`, instead of `&(*data_out)[0]`?